casper/src/main/scala/coop/rchain/casper/util/comm/FairRoundRobinDispatcher.scala

```diff
@@ -0,0 +1,144 @@
+package coop.rchain.casper.util.comm
+
+import scala.collection.immutable.Queue
+
+import cats.Show
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+
+import FairRoundRobinDispatcher._
+import coop.rchain.shared.Log
+
+class FairRoundRobinDispatcher[F[_]: Sync: Log, S: Show, M](
+    filter: M => F[Dispatch],
+    handle: (S, M) => F[Unit],
+    queue: Ref[F, Queue[S]],
+    messages: Ref[F, Map[S, Queue[M]]],
+    retries: Ref[F, Map[S, Int]],
+    skipped: Ref[F, Int],
+    maxSourceQueueSize: Int,
+    giveUpAfterSkipped: Int,
+    dropSourceAfterRetries: Int
+) {
+  assert(maxSourceQueueSize > 0)
+  assert(giveUpAfterSkipped >= 0)
+  assert(dropSourceAfterRetries >= 0)
+
+  def dispatch(source: S, message: M): F[Unit] =
```

Comment:
Is this called synchronously? If the answer is no then I'd be wary of potential races eg. you're in the middle of dropping a source (say just before line 85) while the message arrives from the same source and proceeds to line 51. Then you execute the line 85 and boom

Reply:
Yes, this code is called synchronously. There is a smart constructor that uses a `Semaphore` for concurrent use cases.